tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts

```diff
@@ -0,0 +1,127 @@
+
+import {Chart} from '../chart';
+import {ChartOption} from '../chart_types';
+import {DataSeries} from '../internal_types';
+import {
+  GuestToMainType,
+  InitMessage,
+  MainToGuestEvent,
+  MainToGuestMessage,
+  RendererType,
+} from './worker_chart_types';
+
+self.addEventListener('message', (event: MessageEvent) => {
+  createPortHandler(event.ports[0], event.data as InitMessage);
+});
+
+function createPortHandler(port: MessagePort, initMessage: InitMessage) {
+  const {
+    canvas,
+    devicePixelRatio,
+    dim,
+    rendererType,
+    xScaleType,
+    yScaleType,
+  } = initMessage;
+
+  const lineChartCallbacks = {
+    onDrawEnd: () => {
+      port.postMessage({
+        type: GuestToMainType.ON_REDRAW_END,
+      });
+    },
+  };
+
+  let chartOption: ChartOption;
+  switch (rendererType) {
+    case RendererType.WEBGL:
+      chartOption = {
+        readAndUpdateDomDimensions: RendererType.WEBGL,
+        domDimension: dim,
+        callbacks: lineChartCallbacks,
+        container: canvas,
+        devicePixelRatio,
+        xScaleType,
+        yScaleType,
+      };
+      break;
+    default:
+      throw new RangeError(
+        `Invariant error: cannot have Offscreen chart for renderer type: ${rendererType}`
+      );
+  }
+
+  const lineChart = new Chart(chartOption);
+
+  port.onmessage = function (event: MessageEvent) {
+    const message = event.data as MainToGuestMessage;
+
+    switch (message.type) {
+      case MainToGuestEvent.SERIES_DATA_UPDATE: {
+        const rawData = new Float32Array(message.flattenedSeries);
+        const data: DataSeries[] = [];
+        let rawDataIndex = 0;
+
+        for (const {id, length} of message.idsAndLengths) {
+          const points = [] as Array<{x: number; y: number}>;
+          for (let index = 0; index < length; index++) {
+            points.push({
+              x: rawData[rawDataIndex++],
+              y: rawData[rawDataIndex++],
+            });
+          }
+          data.push({
+            id,
+            points,
+          });
+        }
+
+        lineChart.updateData(data);
+        break;
+      }
+      case MainToGuestEvent.SERIES_METADATA_CHANGED: {
+        lineChart.updateMetadata(message.metadata);
+        break;
+      }
+      case MainToGuestEvent.SERIES_DATA_UPDATE: {
+        break;
+      }
+      case MainToGuestEvent.UPDATE_VIEW_BOX: {
+        lineChart.updateViewBox(message.extent);
+        break;
+      }
+      case MainToGuestEvent.RESIZE: {
+        lineChart.resize(message.dim);
+        break;
+      }
+      case MainToGuestEvent.SCALE_UPDATE: {
+        switch (message.axis) {
+          case 'x':
+            lineChart.setXScaleType(message.scaleType);
+            break;
+          case 'y':
+            lineChart.setYScaleType(message.scaleType);
+            break;
+          default:
+            const _: never = message.axis;
+            throw new RangeError(`Unknown axis: ${_}`);
```

Comment:
Please don’t dereference `_`; it’s confusing, since `_` is supposed to
be a black-hole pattern. If you need to dereference it, just give it a
real name: `const axis`?